File: 380_Insert Delete Get Random O(1).java
// Time complexity: O(1) for each function call.
// Space complexity: O(n)

class RandomizedSet {
    private Map<Integer, Integer> numMap;
    private List<Integer> nums;
    private Random rand;

    public RandomizedSet() {
        this.numMap = new HashMap<>();
        this.nums = new ArrayList<>();
        this.rand = new Random();
    }
    
    public boolean insert(int val) {
        if (numMap.containsKey(val)) return false;
        numMap.put(val, nums.size());
        nums.add(val);
        return true;
    }
    
    public boolean remove(int val) {
        if (!numMap.containsKey(val)) return false;
        int index = numMap.get(val), replaceVal = nums.get(nums.size() - 1);
        nums.set(index, replaceVal);
        nums.remove(nums.size() - 1);
        numMap.put(replaceVal, index);
        numMap.remove(val);
        return true;
    }
    
    public int getRandom() {
        return nums.get(rand.nextInt(nums.size()));
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: 380_Insert Delete Get Random O(1).py
class RandomizedSet:
    # insert/remove O(1): track w/ hashset or dictionary
    # random O(1): fetch from array index
    def __init__(self):
        self.record = {}   # val:idx
        self.nums = []

    def insert(self, val: int) -> bool:
        if val in self.record: return False
        self.record[val] = len(self.nums)
        self.nums.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self.record: return False
        # swap with the tail element, then update & pop tail
        tail = self.nums[-1]
        idx = self.record[val]

        self.nums[idx] = tail
        self.record[tail] = idx

        del self.record[val]
        self.nums.pop()
        return True

    def getRandom(self) -> int:
        idx = random.randint(0, len(self.nums) - 1)
        return self.nums[idx]


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()